### PR TITLE
fix: fix run summary for cached sbt builds

### PR DESCRIPTION
**Pull request by Jorg88**

### CODE REVIEW

File: src/mpyl/run_plan.py [hunk 1]
```diff
@@ -140,36 +140,53 @@ def write_to_pickle_file(self):
...
+
+        summary = "| 👷 Project | 🏗 Build | 🧪 Test | 🚀 Deploy | 🦺 Post-deploy |\n"
+        summary += "| ---------- | :------: | :-----: | :-------: | :------------: |\n"
+
         all_executions = self._get_all_executions()
         if len(all_executions) == 0:
             summary = "Nothing to do 🤷\n"
 
-        for execution in sorted(all_executions, key=operator.attrgetter("name")):
+        unique_executions = {}
```
Comment (ptab) [inline src/mpyl/run_plan.py:175]: What you're actually getting here is the set of all `Project`s in the run plan, in which case it would be better to add this as a method of `RunPlan` that can be unit tested, etc:
```python
def get_all_projects(self, use_full_plan: bool = False) -> set[Project]:
  return { p.project for p in self._get_all_executions(use_full_plan) }
```

Then the code here would become simply:
```python
for project in sorted(
  self.get_all_projects(), key=operator.attrgetter("name")
):
  build_plan = get_icon(project, "build")
  test_plan = get_icon(project, "test")
  deploy_plan = "🚀" if is_execution_in_stage(project, "deploy") else ""
  …
```

**Review by ptab — CHANGES_REQUESTED** [review]

File: src/mpyl/run_plan.py (hunk 1 above)
Comment (Jorg88) [reply src/mpyl/run_plan.py:175]: I can indeed extract it into a function, but your function above is not the same thing. We discussed this at length, but `_get_all_executions` return (for example) discounts twice if its build stage is cached and the test stage is not. It's also not a `list[Project]` that I need, but still a `list[ProjectExecution]`. But I'm using a `dict[str, ProjectExecution]` so I can filter out non-unique names, not matter what the cached status is.

Comment (Jorg88) [reply src/mpyl/run_plan.py:175]: Now that I wrote the above, I don't think there is any value in extracting the logic. It's very specific for this function only

Comment (ptab) [reply src/mpyl/run_plan.py:175]: We did discuss this before, it's clear I'm still struggling to explain my point 😅 Let me try again.

There's nothing particularly wrong with this version of the code, it works and it produces the right output. However, it's needlessly confusing and forces the reader to ask questions they shouldn't have to. 

Starting with a basic example:
```python
{
  "build": [
    ProjectExecution(project="a", cached=True),
    ProjectExecution(project="b", cached=False)
  ], 
  "test": [
    ProjectExecution(project="a", cached=False)
  ]
}
```

With this plan, `unique_executions` will be
```python
unique_executions = [
    ProjectExecution(project="a", cached=True),
    ProjectExecution(project="b", cached=False)
]
```

Then for each of those, we call:
```python
def get_icon(project_execution: ProjectExecution, stage_name: str):
    …    
    if project_execution.project.pipeline == "sbt":
        executions_for_stage = self.get_executions_for_stage_name(stage_name)
        execution_for_stage = next(
            (
                execution_for_stage
                for execution_for_stage in executions_for_stage
                if execution_for_stage.name == project_execution.name
            ),
            None,
        )
```
My immediate first question was: we already have a `ProjectExecution`, so **why do we need to find another one?** And is there really a good way to answer this question without going into a lot of implementation details?

---

Putting the above aside for a moment, let's think about how one would describe `write_to_summary_file()` in plain english. What about "for every **project** in the run plan, print its name and each stage in which it runs and/or is cached"?

Translating this to pseudo-code:
```python
for project in run_plan.get_all_projects():
  name = project.name
  build = run_plan.icon_for(project, "build")
  test = run_plan.icon_for(project, "test")
  deploy = run_plan.icon_for(project, "deploy")

  print "$name | $build | $test | $deploy"  
```

This is very much supported by the fact that, reusing the previous example:
```python
unique_executions = [
  ProjectExecution(project="a", cached=True),
  ProjectExecution(project="b", cached=False)
]
```

**At no point in the entire `write_to_summary_file()` function we use the value of `cached` from these two objects**, which means the above is **functionally equivalent** to:
```python
unique_projects = [
    Project(project="a"),
    Project(project="b")
]
```

I'm really struggling to understand why you're forcing the `write_to_summary_file()` to deal with `ProjectExecution`, when it would be much simpler to make it deal with `Project`.

Comment (Jorg88) [reply src/mpyl/run_plan.py:175]: Ah, I'm following you now. Except that we do use `cached` in the `get_icon` function for the sbt pipeline. Otherwise this thing got be a lot more simple indeed

Comment (ptab) [reply src/mpyl/run_plan.py:175]: Actually we don't, we use the `cached` **of the new project execution we find** (`execution_for_stage`), not the one coming as an input parameter (`project_execution`) 🙂

**Review by ptab — APPROVED** [review]
Awesome, thanks for bearing with me! 💪

**Comment by Jorg88** [comment]
> Awesome, thanks for bearing with me! 💪

No worries, you were actually right. I just didn't fully get what you were trying to say
